canonical/cloud-init
# vmware: Fall back to vmtoolsd if vmware-rpctool errs

Addresses #4436

## Proposed Commit Message


```
vmware: Fallback to vmtoolsd if rpctool errs

This patch udpates the ds-identify script and the
VMware datasource to fall back to using the
vmtoolsd program if vmware-rpctool errors.

Fixes GH-4436
```

## Additional Context

I validated the output and exit code for both commands:

```shell
$ vmtoolsd --cmd "info-get guestinfo.userdata.encoding"; echo $?
gzip+base64
0
```

```shell
$ vmware-rpctool "info-get guestinfo.userdata.encoding"; echo $?
gzip+base64
0
```

```shell
$ vmtoolsd --cmd "info-get guestinfo.userdata.enc"; echo $?
No value found
1
```

```shell
$ vmware-rpctool "info-get guestinfo.userdata.enc"; echo $?
No value found
1
```

## Test Steps

* Locally I verified the OVF datasource with:

    ```shell
    make clean_pyc && PYTHONPATH="$(pwd)" python3 -m pytest -v --log-level=DEBUG tests/unittests/sources/test_ovf.py 
    ```

* Locally I verified the VMware datasource with:

    ```shell
    make clean_pyc && PYTHONPATH="$(pwd)" python3 -m pytest -v --log-level=DEBUG tests/unittests/sources/test_vmware.py 
    ```

* Locally I verified `ds-identify` with:

    ```shell
    make clean_pyc && PYTHONPATH="$(pwd)" python3 -m pytest -v --log-level=DEBUG tests/unittests/test_ds_identify.py
    ```

* I also copied the modified `ds-identify`  and datasources to a VM and ran the following:

    ```shell
      sudo rm -f cloud-init*.log; sudo cloud-init clean; sudo cloud-init init
    ```

    Everything behaved as expected.


* I also replaced `/usr/bin/vmware-rpctool` with:

    ```shell
    #!/bin/sh

    exit 1
    ```

    And ran:

    ```shell
    sudo rm -f cloud-init*.log; sudo cloud-init clean; sudo cloud-init init
    ```

    And `/var/log/cloud-init.log` showed the expected behavior occurred:

    ```log
    2023-09-18 18:29:04,244 - util.py[WARNING]: Failed to get guestinfo value for key metadata: Unexpected error while running command.
    Command: ['/usr/bin/vmware-rpctool', 'info-get guestinfo.metadata']
    Exit code: 1
    Reason: -
    Stdout: 
    Stderr: 
    2023-09-18 18:29:04,244 - util.py[DEBUG]: Failed to get guestinfo value for key metadata: Unexpected error while running command.
    Command: ['/usr/bin/vmware-rpctool', 'info-get guestinfo.metadata']
    Exit code: 1
    Reason: -
    Stdout: 
    Stderr: 
    Traceback (most recent call last):
      File "/usr/lib/python3.11/site-packages/cloudinit/sources/DataSourceVMware.py", line 551, in guestinfo_get_value
        (stdout, stderr) = subp(args)
                          ^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/cloudinit/subp.py", line 335, in subp
        raise ProcessExecutionError(
    cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
    Command: ['/usr/bin/vmware-rpctool', 'info-get guestinfo.metadata']
    Exit code: 1
    Reason: -
    Stdout: 
    Stderr: 
    2023-09-18 18:29:04,246 - DataSourceVMware.py[DEBUG]: Getting guestinfo value for key metadata with /usr/bin/vmtoolsd
    2023-09-18 18:29:04,246 - subp.py[DEBUG]: Running command ['/usr/bin/vmtoolsd', '--cmd', 'info-get guestinfo.metadata'] with allowed return codes [0] (shell=False, capture=True)
    2023-09-18 18:29:04,295 - DataSourceVMware.py[DEBUG]: Getting guestinfo value for key metadata.encoding with /usr/bin/vmtoolsd
    2023-09-18 18:29:04,295 - subp.py[DEBUG]: Running command ['/usr/bin/vmtoolsd', '--cmd', 'info-get guestinfo.metadata.encoding'] with allowed return codes [0] (shell=False, capture=True)
    2023-09-18 18:29:04,329 - DataSourceVMware.py[DEBUG]: Getting encoded data for key=guestinfo.metadata, enc=gzip+base64
    2023-09-18 18:29:04,330 - DataSourceVMware.py[DEBUG]: Decoding gzip+base64 format guestinfo.metadata
    2023-09-18 18:29:04,330 - DataSourceVMware.py[DEBUG]: Getting guestinfo value for key userdata with /usr/bin/vmtoolsd
    2023-09-18 18:29:04,330 - subp.py[DEBUG]: Running command ['/usr/bin/vmtoolsd', '--cmd', 'info-get guestinfo.userdata'] with allowed return codes [0] (shell=False, capture=True)
    2023-09-18 18:29:04,373 - DataSourceVMware.py[DEBUG]: Getting guestinfo value for key userdata.encoding with /usr/bin/vmtoolsd
    2023-09-18 18:29:04,373 - subp.py[DEBUG]: Running command ['/usr/bin/vmtoolsd', '--cmd', 'info-get guestinfo.userdata.encoding'] with allowed return codes [0] (shell=False, capture=True)
    2023-09-18 18:29:04,411 - DataSourceVMware.py[DEBUG]: Getting encoded data for key=guestinfo.userdata, enc=gzip+base64
    2023-09-18 18:29:04,411 - DataSourceVMware.py[DEBUG]: Decoding gzip+base64 format guestinfo.userdata
    2023-09-18 18:29:04,411 - DataSourceVMware.py[DEBUG]: Getting guestinfo value for key vendordata with /usr/bin/vmtoolsd
    ```

* With `/usr/bin/vmware-rpctool` still replaced, I validated `ds-identify` by running `sudo rm /var/run/cloud-init/.ds-identify.result /var/run/cloud-init/ds-identify.log` and then running `sudo /usr/lib/cloud-init/ds-identify`. The resulting log fail showed the VMware datasource was still detected:

    ```shell
    $ sudo stat /var/run/cloud-init/ds-identify.log 
      File: /var/run/cloud-init/ds-identify.log
      Size: 1360      	Blocks: 8          IO Block: 4096   regular file
    Device: 0,23	Inode: 2318        Links: 1
    Access: (0640/-rw-r-----)  Uid: (    0/    root)   Gid: (    0/    root)
    Access: 2023-09-18 19:54:51.286717602 +0000
    Modify: 2023-09-18 19:54:48.650490201 +0000
    Change: 2023-09-18 19:54:48.650490201 +0000
     Birth: 2023-09-18 19:54:48.498477087 +0000
    ```
    
    ```
    [up 1725229.52s] ds-identify 
    policy loaded: mode=search report=false found=all maybe=all notfound=disabled
    /etc/cloud/cloud.cfg set datasource_list: ['NoCloud', 'ConfigDrive', 'OpenStack', 'VMware', None]
    DMI_PRODUCT_NAME=VMware20,1
    DMI_SYS_VENDOR=VMware, Inc.
    DMI_PRODUCT_SERIAL=VMware-42 1a b5 b3 64 97 90 e6-35 7b d8 97 66 df c3 d0
    DMI_PRODUCT_UUID=b3b51a42-9764-e690-357b-d89766dfc3d0
    PID_1_PRODUCT_NAME=unavailable
    DMI_CHASSIS_ASSET_TAG=No Asset Tag
    DMI_BOARD_NAME=440BX Desktop Reference Platform
    FS_LABELS=
    ISO9660_DEVS=
    KERNEL_CMDLINE=BOOT_IMAGE=/boot/vmlinuz-6.1.10-10.ph5-esx root=PARTUUID=c724cded-cf33-4d1c-b510-59478c322f84 init=/lib/systemd/systemd rcupdate.rcu_expedited=1 rw systemd.show_status=0 quiet noreplace-smp cpu_init_udelay=0 net.ifnames=0 plymouth.enable=0 systemd.unified_cgroup_hierarchy=yes
    VIRT=vmware
    UNAME_KERNEL_NAME=Linux
    UNAME_KERNEL_RELEASE=6.1.10-10.ph5-esx
    UNAME_KERNEL_VERSION=#1-photon SMP Mon Apr 24 22:51:08 UTC 2023
    UNAME_MACHINE=x86_64
    UNAME_NODENAME=photon5-w-tpm
    UNAME_OPERATING_SYSTEM=GNU/Linux
    DSNAME=
    DSLIST=NoCloud ConfigDrive OpenStack VMware None
    MODE=search
    ON_FOUND=all
    ON_MAYBE=all
    ON_NOTFOUND=disabled
    pid=24540 ppid=24539
    is_container=false
    is_ds_enabled(IBMCloud) = false.
    is_ds_enabled(IBMCloud) = false.
    check for 'VMware' returned found
    Found single datasource: VMware
    [up 1725229.68s] returning 0
    ``` 

## Checklist:
